PRCS/prc-4.md

```diff
@@ -0,0 +1,151 @@
+
+    /// @notice Returns the seller's current `orderId` (index that their new sell order will be mapped to).
+    function getSellerOrderId(address seller) external view returns (uint256);
+
+    /// @notice Returns the Order struct information about an order identified by the combination `<seller, orderId>`.
+    function getOrder(address seller, uint256 orderId) external view returns (Order memory);
+
+    /// @notice Creates a sell order with incremental seller-specific `orderId` for the specified `assetAmount` at specified `pricePerAsset`.
+    /// @dev The order information is saved in a nested mapping `seller address -> orderId -> Order`.
+    /// MUST emit `OrderCreated` event.
+    function createSellOrder(uint256 assetAmount, uint256 pricePerAsset) external;
+
+    /// @notice Consecutively fills an array of orders identified by the combination `<seller, orderId>`,
+    /// by providing an exact amount of ETH and requesting a specific minimum amount of asset to receive.
+    /// @dev Transfers portions of msg.value to the orders' sellers according to the price.
+    /// The sum of asset amounts of filled orders MUST be at least `minimumAsset`.
+    /// If msg.value is more than the sum of orders' prices, it MUST refund the excess back to msg.sender.
+    /// An order whose `cancelled` parameter has value `true` MUST NOT be filled.
+    /// MUST change the `assetAmount` parameter for the specified order according to how much of it was filled.
+    /// MUST emit `OrderFilled` event for each order accordingly.
+    function fillOrdersExactEth(uint256 minimumAsset, address payable[] memory sellers, uint256[] memory orderIds)
+        external
+        payable;
+
+    /// @notice Consecutively fills an array of orders identified by the combination `<seller, orderId>`,
+    /// by providing a possibly surplus amount of ETH and requesting an exact amount of asset to receive.
+    /// @dev Transfers portions of msg.value to the orders' sellers according to the price.
+    /// The sum of asset amounts of filled orders MUST be exactly `assetAmount`. Excess ETH MUST be returned back to `msg.sender`.
+    /// An order whose `cancelled` parameter has value `true` MUST NOT be filled.
+    /// MUST change the `assetAmount` parameter for the specified order according to how much of it was filled.
+    /// MUST emit `OrderFilled` event for each order accordingly.
+    /// If msg.value is more than the sum of orders' prices, it MUST refund the difference back to msg.sender.
+    function fillOrdersExactAsset(uint256 assetAmount, address payable[] memory sellers, uint256[] memory orderIds)
+        external
+        payable;
+
+    /// @notice Cancels the sell order identified by combination `<msg.sender, orderId>`, making it unfillable.
+    /// @dev MUST change the `cancelled` parameter for the specified order to `true`.
+    /// MUST emit `OrderCancelled` event.
+    function cancelSellOrder(uint256 orderId) external;
```

Comment:
From a gane perspective, when you see a cancel event emitted, how do you know how much gold to unlock for which user?

orderId here would either have to be a gobal id so we can look it up or we need more context